Update dependencies and Python version

Grimpy101 opened the pull request:
Changes and additions:
- Update to Python 3.14 (although back-compatibility to Python 3.8 has been tested as well)
  - Moved from comment-based type annotation to Python 3 type hints
  - Changed some parts with unicode type to bytes (Pyscopg2Backend)
  - Removed absolute_import
  - Adjusted tests (list comparison was changed to set comparison, built-in testing dependency is used)
  - setup.py and setup.cfg were removed in favour of pyproject.toml
- Update dependencies (Flask, psycopg2)
  - Bumped Flask version
  - Switched from deprecated flask._app_ctx_stack to flask.g (PhpBB3)
  - Moved from werkzeug.contrib.cache to cachelib library
  - In tests, set_cookie(...) function arguments were reordered, and server address was specified in client config

Notes:
- .travis.yml was updated, but the new configuration was not tested
- All files were structured in accordance with flake8 and mypy configuration

Grimpy101 committed:
Update dependencies

Grimpy101 committed:
Update BaseBackend class

Grimpy101 committed:
Update UserAcl

Grimpy101 committed:
Update UserAcl

Grimpy101 committed:
Update Python version in mypy config

Grimpy101 committed:
Modify BaseBackend.execute method to improve typing

Grimpy101 committed:
Update Psycopg2Backend

Grimpy101 committed:
Remove absolute_import

Grimpy101 committed:
Bump Python version in configs

Grimpy101 committed:
Update imports in psycopg2.py, extensions.py, and sessions.py

Grimpy101 committed:
Update PhpBB3, reassign phpbb3_backend from flask._app_ctx_stack to flask.g

Grimpy101 committed:
Update sessions.py, add error handling in extension.py

Grimpy101 committed:
Add comments

Grimpy101 committed:
Fixing tests WIP

Grimpy101 committed:
Update requirements, adjust tests for this Flask version

Grimpy101 committed:
Format tests

Grimpy101 committed:
Update package building

Grimpy101 committed:
Switch ordering of classes to make types Python 3.8 compatible, update imports

MasterMind2k commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-Flask>=0.10
-typing
+psycopg2>=2.9.11
+Flask>=3.1.2
```
Dependencies should be more "inclusive" regarding dependencies. Is latestflask really needed?

Grimpy101 replied:
Sure, we can probably lower the Flask version to 2.3.0 (this is where support for Python 3.7 was dropped and most breaking changes were introduced). For psycopg2, version 2.9.10 would likely suffice (that's where the support for Python 3.7 was dropped as well), although I need to check if it is compatible with Python 3.14 ...

MasterMind2k replied:
We can make intermediate version to have 3.8 version supported.

Afaik, only podnapisi.net uses this library :) .

MasterMind2k commented on flask_phpbb3/backends/psycopg2.py:
```diff
@@ -1,7 +1,7 @@
 from __future__ import absolute_import
 
 import json
-import typing
```
Please use module imports.

In python, non-modules (function, classes, etc.) get new reference and are harder to mock. I understad, `typing` will never be mocked, but it has to be imported in same way as everything else.

Grimpy101 replied:
Noted :+1:
I will correct it.

MasterMind2k commented on flask_phpbb3/backends/psycopg2.py:
```diff
@@ -106,13 +103,12 @@ def _prepare_statements(self):
         """
         # Setters for custom fields
-        custom_fields = self._config.get('CUSTOM_USER_FIELDS', [])
+        custom_fields: List[Any] = self._config.get('CUSTOM_USER_FIELDS', [])
```
We can safely assume this is list of strings.

MasterMind2k commented on flask_phpbb3/extension.py:
```diff
@@ -101,154 +98,164 @@ def _create_backend(
-            return backend
-        raise AttributeError('No context available')
+        backend: Optional[flask_phpbb3.backends.base.BaseBackend] =\
+            flask.g.get('phpbb3_backend', None)
```
Could you describe why this change?

Grimpy101 replied:
In Flask 2.3, flask._app_ctx_stack [was removed](https://flask.palletsprojects.com/en/stable/changes/#version-2-3-0). The preferred way, as I understand it, is to use Flask's g object to store global data (as noted in [this GitHub post](https://github.com/nickw444/flask-ldap3-login/pull/115)).

MasterMind2k replied:
Great explanation! Something that has place in git commit.

MasterMind2k commented on flask_phpbb3/sessions.py:
```diff
@@ -8,66 +10,56 @@
-class PhpBB3Session(dict, flask.sessions.SessionMixin):
-    def __init__(self):
-        # type: () -> None
+class PhpBB3Session(Dict[str, Any], flask.sessions.SessionMixin):
```
Why inherit from `typing.Dict`?

Grimpy101 replied:
Fair, I can change it back, I just wanted more granular typing (dict with string keys in this case). But as far as I know, typing.Dict behaves as dict when inherited.

MasterMind2k commented on pyproject.toml:
Very nice!!!

MasterMind2k commented on .travis.yml:
```diff
@@ -4,15 +4,10 @@ services:
-    else
-      pip install -r requirements/dev-3.txt
-    fi
+    pip install -r requirements/dev-3.txt
```
Could you also prepare environment management with `uv`?

Grimpy101 replied:
I will look into it :)

MasterMind2k commented on flask_phpbb3/backends/base.py:
```diff
@@ -1,83 +1,12 @@
 ACL_OPTIONS_CACHE_TTL: int = 3600 * 1
 
 
-class BaseBackend:
```
That's a big commit change for a python 3.8 compatibility :) . It makes review harder.

Grimpy101 replied:
Sorry, I'll try to keep these to a minimum and explain them better. If I recall correctly, BaseBackend was only moved below UserAcl so that the type hints in UserAcl could recognize it.

MasterMind2k replied:
No worries, you are doing great! Maybe give a [Tidy First?](https://www.oreilly.com/library/view/tidy-first/9781098151232/) book a try.

Also, this is a great oportuity to try out `git rebase -i` and master "writing commits". Something that helped me was https://cbea.ms/git-commit/

Grimpy101 replied:
Great, I will have a look at the book! I tried rebasing, so now I have a bit of an idea how to modify commits :)

MasterMind2k reviewed (changes requested):
Awesome!

A few things to consider:
* Write commits as a story, concentrate to write_reason of change_ into commit message (answer the question why). The answer to question What is answered in change itsef
* Be more graceus with dependencies for packages, limit the lowest allowed version of flask and to next major release (hoping flask follows semver)

And some additional wishes:
* switch from `flake8` to `ruff`
* introduce `black` formatter
* Introduce `uv` for handling dependencies
* Go through coerage report and increase test coverage

Thank you for help! Very much appreciated.

Grimpy101 committed:
Switch to uv package management

Grimpy101 committed:
Unify module imports

Grimpy101 committed:
Specify definite typing based on context, change inharitance to python in-built types

Grimpy101 committed:
Format files with Black formatter

Grimpy101 commented:
I have tried to address all feedback, and I've added Black formatter and ruff, and introduced uv for package management. Maybe I'll also extend test coverage when I have time.
